### PR TITLE
[Transforms] Fixes for ShorthandPropertyAssignments and exported namespaces

### LINKED ISSUE #7862: [Transforms] Incorrect emit for ShorthandPropertyAssignment for an exported value
**Expected**:

```js
  var aa;
    (_a = { a: a, b: b }, a = _a.a, b = _a.b, _a);
    (_b = { b: b, a: a }, a = _b.a, b = _b.b, _b);
```

**Actual**:

```js
  var aa;
    (_a = { a, b }, a = _a.a, b = _a.b, _a);
    (_b = { b, a }, a = _b.a, b = _b.b, _b);
```

Failed tests:
*  tests/cases/conformance/es6/destructuring/declarationsAndAssignments.ts
* objectLiteralShorthandPropertiesWithModuleES6.js
* giant.js

The last two are failures from namespace exports, not module exports.

### PR DESCRIPTION
**Pull request by rbuckton**
This change revises how JIT substitution works in the printer. Previously we had two hooks for substitution on `TransformationContext`: `identifierSubstitution` and `expressionSubstitution`. Now, we instead have a single hook, `onSubstituteNode`, which adds an additional `isExpression` parameter that will be provided a value indicating whether the node to be substituted is expected to be used as an expression.

This change allows us to easily add substitutions for a node that is neither an identifier nor an expression, such as a `ShorthandPropertyAssignment`.

Fixes #7862.

Fixes the following tests:
* tests/cases/conformance/es6/destructuring/declarationsAndAssignments.ts
* tests/cases/conformance/es6/shorthandPropertyAssignment/objectLiteralShorthandPropertiesWithModuleES6.ts
* tests/cases/conformance/internalModules/moduleDeclarations/invalidInstantiatedModule.ts
* tests/cases/compiler/giant.ts
* tests/cases/compiler/privacyImportParseErrors.ts
* tests/cases/compiler/shorthand-property-es6-amd.ts
* tests/cases/compiler/shorthandPropertyAssignmentInES6Module.ts

### CODE REVIEW

File: src/compiler/printer.ts [hunk 1]
```diff
@@ -2216,12 +2173,16 @@ const _super = (function (geti, seti) {
                 }
             }
 
-            function tryEmitSubstitute(node: Node, substitution: (node: Node) => Node) {
-                if (substitution && (getNodeEmitFlags(node) & NodeEmitFlags.NoSubstitution) === 0) {
-                    const substitute = substitution(node);
+            function tryEmitSubstitute(node: Node, emitNode: (node: Node) => void, isExpression: boolean) {
+                if ((<any>node).text === "localizedDiagnosticMessages" && isExpression) {
```
Comment (mhegazy) [inline src/compiler/printer.ts]: do not think you need this.

File: src/compiler/transformers/destructuring.ts [hunk 2]
```diff
@@ -397,16 +397,22 @@ namespace ts {
...
      * @param emitTempVariableAssignment A callback used to emit a temporary variable.
+     * @param visitor An optional callback used to visit the value.
      */
     function ensureIdentifier(
         value: Expression,
         reuseIdentifierExpressions: boolean,
         location: TextRange,
-        emitTempVariableAssignment: (value: Expression, location: TextRange) => Identifier) {
+        emitTempVariableAssignment: (value: Expression, location: TextRange) => Identifier,
+        visitor?: (node: Node) => VisitResult<Node>) {
```
Comment (sandersn) [inline src/compiler/transformers/destructuring.ts]: tabbing is weird here

**Comment by sandersn** [comment]
:+1:

File: src/compiler/printer.ts (hunk 1 above)
Comment (rbuckton) [inline src/compiler/printer.ts]: Ah, thanks. Left-over from trying to fix an issue.

**Pull request merged**